src/wallet/gtest/test_deadlock.cpp

```diff
@@ -0,0 +1,80 @@
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <sodium.h>
+
+#include <boost/filesystem.hpp>
+#include <thread>
+#include <atomic>
+
+#include "zcash/JoinSplit.hpp"
+#include "wallet/wallet.h"
+#include "utiltest.h"
+
+using namespace libzcash;
+
+extern ZCJoinSplit *params;
+
+void write_db(CWallet &wallet, CBlock &block, SpendingKey &sk, std::atomic_int &finish)
+{
+    for (int i = 0; i < 1000; i++) {
+        auto wtx = GetValidReceive(*params, sk, 10, true);
+        wallet.SyncTransaction(wtx, &block);
+    }
+
+    finish++;
+}
+
+void write_block(CWallet &walletdb, std::atomic_int &finish) {
+    for (int i = 0; i < 100; i++) {
+        CBlockLocator loc;
+        walletdb.SetBestChain(loc);
+    }
+
+    finish++;
+}
+
+TEST(deadlock_test, deadlock) {
+    //set tmp db folder
+    boost::filesystem::path pathTemp = boost::filesystem::temp_directory_path()
+            / boost::filesystem::unique_path();
+    boost::filesystem::create_directories(pathTemp);
+    mapArgs["-datadir"] = pathTemp.string();
+
+    //create wallet
+    SelectParams(CBaseChainParams::TESTNET);
+    bool fFirstRun = true;
+    CWallet walletMain("deadlock_ut_wallet.dat");
+    DBErrors nLoadWalletRet = walletMain.LoadWallet(fFirstRun);
+
+    auto sk = libzcash::SpendingKey::random();
+    walletMain.AddSpendingKey(sk);
+
+    //create block
+    CBlock block;
```

Comment:
Maybe the obj can be created in the function thread (instead of passing in the thread creation)